### DISCUSSION TITLE
can't call computeOffDiagJacobian in a Kernel?

### DISCUSSION THREAD

**Discussion by bylore** [idaholab/moose#24268] [Q&A General] [answered]
My kernel inherits Kernel and I override thest three functions
```
  virtual Real computeQpResidual() override;
  virtual Real computeQpJacobian() override;
  virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;
```
computeQpResidual() and computeQpJacobian() work well, but computeQpOffDiagJacobian() doesn't seem to be getting called. 

Could you guys give me any help?

**Comment by bylore** [comment]
`class TermiteStressCoupleElectricFieldOffDiag : public Kernel
{
public:
  static InputParameters validParams();

  TermiteStressCoupleElectricFieldOffDiag(const InputParameters & parameters);

protected:
  virtual Real computeQpResidual() override;

  virtual Real computeQpJacobian() override;
  
  virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;

  /// Base name of the material system that this kernel applies to
  const std::string _base_name;

  /// An integer corresponding to the direction this kernel acts in
  const unsigned int _component;

  const unsigned int _ep_var; //index of the coupled variable electric potential
  const MaterialProperty<RankThreeTensor> & _d_stress_d_efield;
};`

`Real
TermiteStressCoupleElectricFieldOffDiag::computeQpResidual()
{
  //std::cout << "222" << std::endl;
  return 0.0;
}

Real
TermiteStressCoupleElectricFieldOffDiag::computeQpJacobian()
{
  //std::cout << "333" << std::endl;
  return 0.0;
}

Real
TermiteStressCoupleElectricFieldOffDiag::computeQpOffDiagJacobian(unsigned int jvar)
{
  std::cout << "444" << std::endl;
  if (jvar == _ep_var)
  {
    Real val = 0.0;
    for (unsigned int j=0; j<Moose::dim; ++j)
      for(unsigned int k=0; k<Moose::dim; ++k)
      {
        //std::cout << _d_stress_d_efield[_qp](_component, j, k) << std::endl;
        val -= _d_stress_d_efield[_qp](_component, j, k) * _grad_test[_j][_qp](k) * _grad_test[_i][_qp](j);
      }
    return val;
  }

  // Returns if coupled variable is not electric potential
  return 0.0;
}`

**Reply by bylore** [reply]
"444" can not be printed in my test simulation.

**Reply by GiudGiud** [reply]
What kind of preconditioning are you using?

how many non linear variables do you have?

**Reply by YaqiWang** [reply, marked as answer]
By default MOOSE only assembles the variable-wise block diagonal Jacobian, thus this function is not called. If you switch `Executioner/solve_type` to NEWTON, or add a preconditioner block with SMP and turn the full coupling on, you will see this function being called.

**Reply by bylore** [reply]
@GiudGiud, hi, my confusion has been pointed out by @YaqiWang. Thank you guys for the reply and answer!
